test/validation/api/ml/ml.c

```diff
@@ -0,0 +1,485 @@
+	odp_ml_compl_pool_param_init(&pool_param);
+	pool_param.uarea_init.init_fn = init_event_uarea;
+	pool_param.uarea_init.args = &data;
+	pool_param.num = num;
+	pool_param.uarea_size = 1;
+	pool = odp_ml_compl_pool_create(NULL, &pool_param);
+
+	CU_ASSERT_FATAL(pool != ODP_POOL_INVALID);
+	CU_ASSERT(data.count == num);
+
+	for (i = 0; i < num; i++) {
+		CU_ASSERT(data.mark[i] == 1);
+
+		compl_evs[i] = odp_ml_compl_alloc(pool);
+
+		CU_ASSERT(compl_evs[i] != ODP_ML_COMPL_INVALID);
+
+		if (compl_evs[i] == ODP_ML_COMPL_INVALID)
+			break;
+
+		uarea = odp_ml_compl_user_area(compl_evs[i]);
+
+		CU_ASSERT(*uarea == UAREA);
+	}
+
+	for (uint32_t j = 0; j < i; j++)
+		odp_ml_compl_free(compl_evs[j]);
+
+	odp_pool_destroy(pool);
+}
+
+static void test_ml_fp32_to_uint8(void)
+{
+	uint8_t u8[8];
+	float fp[8] = {-20.f, -16.4f, -14.6f, -12.5f, 0, 31.4f, 80.f, 96.3f};
+	uint8_t expected[8] = {0, 0, 4, 10, 43, 127, 255, 255};
+
+	float scale = 0.3746f;
+	uint8_t zero_point = 43;
+
+	odp_ml_fp32_to_uint8(u8, fp, 8, scale, zero_point);
+	for (uint32_t i = 0; i < 8; i++)
+		CU_ASSERT_EQUAL(u8[i], expected[i]);
+}
+
+static void test_ml_fp32_from_uint8(void)
+{
+	float fp[4];
+	float scale = 0.4f;
+	uint8_t zero_point = 43;
+	uint8_t u8[4] = {0, 43, 145, 255};
+	float expected[4] = {-17.2f, 0.0f, 40.8f, 84.8f};
+
+	odp_ml_fp32_from_uint8(fp, u8, 4, scale, zero_point);
+	for (uint32_t i = 0; i < 4; i++)
+		CU_ASSERT_EQUAL(fp[i], expected[i]);
+}
+
+static void test_ml_fp32_to_int8(void)
+{
+	int8_t i8[5];
+	float scale = 0.0223f;
+	int8_t zero_point = 0;
+	float fp32[5] = {-3.4f, -2.5f, 0, 1.4f, 2.9f};
+	int8_t i8_expected[5] = {-127, -112, 0, 63, 127};
+
+	odp_ml_fp32_to_int8(i8, fp32, 5, scale, zero_point);
+
+	for (uint32_t i = 0; i < 5; i++)
+		CU_ASSERT_EQUAL(i8[i], i8_expected[i]);
+}
+
+static void test_ml_fp32_to_int8_positive_zp(void)
+{
+	int8_t i8[6];
+	float scale = 0.0223f;
+	int8_t zero_point = 56;
+	float fp32[6] = {-4.1f, -3.4f, -2.5f, 0, 1.4f, 2.9f};
+	int8_t i8_expected[6] = {-127, -96, -56, 56, 119, 127};
+
+	odp_ml_fp32_to_int8(i8, fp32, 6, scale, zero_point);
+
+	for (uint32_t i = 0; i < 6; i++)
+		CU_ASSERT_EQUAL(i8[i], i8_expected[i]);
+}
+
+static void test_ml_fp32_to_int8_negative_zp(void)
+{
+	int8_t i8[6];
+	float scale = 0.0223f;
+	int8_t zero_point = -56;
+	float fp32[6] = {-3.4f, -2.5f, 0, 1.4f, 2.9f, 4.1f};
+	int8_t i8_expected[6] = {-127, -127, -56, 7, 74, 127};
+
+	odp_ml_fp32_to_int8(i8, fp32, 6, scale, zero_point);
+
+	for (uint32_t i = 0; i < 6; i++)
+		CU_ASSERT_EQUAL(i8[i], i8_expected[i]);
+}
+
+static void test_ml_fp32_from_int8(void)
+{
+	float fp32[6];
+	float scale = 0.05f;
+	int8_t zero_point = 56;
+	int8_t i8[6] = {-128, 46, 0, 56, 85, 127};
+	float fp32_expected[6] = {-9.2f, -0.5f, -2.8f, 0.0f, 1.45f, 3.55f};
+
+	odp_ml_fp32_from_int8(fp32, i8, 6, scale, zero_point);
+
+	for (uint32_t i = 0; i < 6; i++)
+		CU_ASSERT_EQUAL(fp32[i], fp32_expected[i]);
+}
+
+odp_testinfo_t ml_suite[] = {
```

Comment:
Tests are missing for `odp_ml_fp32_to_fp16()` and `odp_ml_fp32_from_fp16`.

Reply:
The tests would have to be in linux-gen validation. The implementation is adapted from DPDK, so what kind of tests would make sense?

Reply:
The tests belong here in API level.  For example the test(s) could utilize normal floats (float -> fp16 -> float) and allow some variation (e.g. 1%) in the results.

Reply:
I'm not entirely convinced about its usefulness, but anyway test added in v3.